scheduler-pool/src/lib.rs

```diff
@@ -0,0 +1,503 @@
+            ScheduleExecutionArg, SchedulerId, SchedulingContext, WaitReason,
+            WithTransactionAndIndex,
+        },
+        prioritization_fee_cache::PrioritizationFeeCache,
+        vote_sender_types::ReplayVoteSender,
+    },
+    solana_scheduler::{SchedulingMode, WithSchedulingMode},
+    solana_sdk::transaction::Result,
+    std::{
+        fmt::Debug,
+        marker::PhantomData,
+        sync::{Arc, Mutex, Weak},
+    },
+};
+
+// SchedulerPool must be accessed via dyn by solana-runtime code, because of its internal fields'
+// types aren't available there...
+#[derive(Debug)]
+pub struct SchedulerPool {
+    schedulers: Mutex<Vec<InstalledSchedulerBox>>,
+    log_messages_bytes_limit: Option<usize>,
+    transaction_status_sender: Option<TransactionStatusSender>,
+    replay_vote_sender: Option<ReplayVoteSender>,
+    prioritization_fee_cache: Arc<PrioritizationFeeCache>,
+    weak_self: Weak<SchedulerPool>,
+}
+
+impl SchedulerPool {
+    pub fn new(
+        log_messages_bytes_limit: Option<usize>,
+        transaction_status_sender: Option<TransactionStatusSender>,
+        replay_vote_sender: Option<ReplayVoteSender>,
+        prioritization_fee_cache: Arc<PrioritizationFeeCache>,
+    ) -> Arc<Self> {
+        Arc::new_cyclic(|weak_self| Self {
+            schedulers: Mutex::<Vec<InstalledSchedulerBox>>::default(),
+            log_messages_bytes_limit,
+            transaction_status_sender,
+            replay_vote_sender,
+            prioritization_fee_cache,
+            weak_self: weak_self.clone(),
+        })
+    }
+
+    pub fn new_dyn(
+        log_messages_bytes_limit: Option<usize>,
+        transaction_status_sender: Option<TransactionStatusSender>,
+        replay_vote_sender: Option<ReplayVoteSender>,
+        prioritization_fee_cache: Arc<PrioritizationFeeCache>,
+    ) -> InstalledSchedulerPoolArc {
+        Self::new(
+            log_messages_bytes_limit,
+            transaction_status_sender,
+            replay_vote_sender,
+            prioritization_fee_cache,
+        )
+    }
+
+    pub fn self_arc(&self) -> Arc<Self> {
+        self.weak_self
+            .upgrade()
+            .expect("self-referencing Arc-ed pool")
+    }
+}
+
+impl InstalledSchedulerPool for SchedulerPool {
+    fn take_from_pool(&self, context: SchedulingContext) -> InstalledSchedulerBox {
+        let mut schedulers = self.schedulers.lock().expect("not poisoned");
+        // pop is intentional for filo, expecting relatively warmed-up scheduler due to having been
+        // returned recently
+        let maybe_scheduler = schedulers.pop();
+        if let Some(mut scheduler) = maybe_scheduler {
+            scheduler.replace_context(context);
+            scheduler
+        } else {
+            Box::new(PooledScheduler::<_, DefaultScheduleExecutionArg>::spawn(
+                self.self_arc(),
+                context,
+                DefaultTransactionHandler,
+            ))
+        }
+    }
```

Comment:
It seems that `self_arc()` and `self.weak_self` are only necessary in `take_from_pool()`, as the later needs `self` to be an `Arc`.

Rust allows other types than `Self` to be used for the `self` argument, including `Arc<Self>`.
Exactly for cases, when a method needs to rely on the additional properties provided by the wrapper type.

Considering that `new()` is already returning `Arc<Self>` it is possible that callers of `take_from_pool()` are already operating on an `Arc<SchedulerPool>`.
In which case, have you considered changing `take_from_pool()` type to this:
```rust
fn take_from_pool(self: Arc<Self>, context: SchedulingContext) -> InstalledSchedulerBox {
```
I wonder if this would allow you to remove the self reference.

Reply:
you're right but `InstalledSchedulerPool` must be object-safe.. so, it can't take `self: Arc<Self>`...

Reply:
`Arc<Self>` is object safe: https://doc.rust-lang.org/reference/items/traits.html#object-safety

This is one of the things that Rust can do that C++ can not.

Reply:
> `Arc<Self>` is object safe: https://doc.rust-lang.org/reference/items/traits.html#object-safety

yeah, i know. pardon for my previous terse reply... I'd like to avoid to take _owned_ Arc<Self> as it necessitates `.clone()` at every call site: d6487a435f2
and passed `Arc<Self>` is usually dropped for nothing unless the scheduler pool isn't populated yet (which occurs only at the initial use of it).
so, I'd like to use `&Arc<Self>`, but it isn't supported by rust as this is the nested receiver case for the object safe trait.

I know these wasted clones are of rather outside any hot codepath. so, the clone overhead isn't not big deal.

Alternatively, i could `impl InstalledSchedulerPool for Arc<SchedulerPool>`, but i can't do this directly due to the coherence rule, forcing a newtype pattern, which brings another whole set of drawbacks/complications (in fact, this pr was _simplified_ from the new type approach...)

all in all, i'd prefer self referencing Weak inside a struct.

Reply:
Thank you for the explanation.
Considering how tricky this is, consider adding this explanation as a comment in the code.
Otherwise, someone like me reading this code, outside of the context of this PR, could be confused.  And they may even want to remove `Weak`, breaking your carefully crafted optimization.

Reply:
> Thank you for the explanation. ... like me reading this code, outside of the context of this PR ...

no problem. really thanks for jumping on this mega pr as outsider. :) it helps to reveal these unspoken assumptions.

> Considering how tricky this is, consider adding this explanation as a comment in the code. Otherwise, someone ..., could be confused. And they may even want to remove `Weak`, breaking your carefully crafted optimization.

done a241a8b5d4b